src/dynamics/joint.rs

```diff
@@ -1,9 +1,73 @@
-use crate::dynamics::GenericJoint;
 use bevy::prelude::*;
 use rapier::dynamics::{ImpulseJointHandle, MultibodyJointHandle};
 
 pub use rapier::dynamics::{JointAxesMask, JointAxis, MotorModel};
 
+use super::{FixedJoint, GenericJoint, PrismaticJoint, RevoluteJoint, RopeJoint, SpringJoint};
+
+#[cfg(feature = "dim3")]
+use super::SphericalJoint;
+
+/// Wrapper enum over a specific joint.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum JointDescription {
+    /// See [`FixedJoint`]
+    FixedJoint(FixedJoint),
+    /// See [`GenericJoint`]
+    GenericJoint(GenericJoint),
+    /// See [`PrismaticJoint`]
+    PrismaticJoint(PrismaticJoint),
+    /// See [`RevoluteJoint`]
+    RevoluteJoint(RevoluteJoint),
+    /// See [`RopeJoint`]
+    RopeJoint(RopeJoint),
+    /// See [`SphericalJoint`]
+    #[cfg(feature = "dim3")]
+    SphericalJoint(SphericalJoint),
+    /// See [`SpringJoint`]
+    SpringJoint(SpringJoint),
+}
+
+impl JointDescription {
+    /// The underlying generic joint.
+    pub fn generic_joint(&self) -> &GenericJoint {
+        match self {
+            JointDescription::FixedJoint(j) => &j.data,
+            JointDescription::GenericJoint(j) => j,
+            JointDescription::PrismaticJoint(j) => &j.data,
+            JointDescription::RevoluteJoint(j) => j.data(),
+            JointDescription::RopeJoint(j) => j.data(),
+            #[cfg(feature = "dim3")]
+            JointDescription::SphericalJoint(j) => j.data(),
+            JointDescription::SpringJoint(j) => j.data(),
+        }
+    }
+    /// The underlying generic joint.
+    pub fn generic_joint_mut(&mut self) -> &mut GenericJoint {
+        match self {
+            JointDescription::FixedJoint(ref mut j) => &mut j.data,
+            JointDescription::GenericJoint(ref mut j) => j,
+            JointDescription::PrismaticJoint(ref mut j) => &mut j.data,
+            JointDescription::RevoluteJoint(ref mut j) => &mut j.data,
+            JointDescription::RopeJoint(ref mut j) => &mut j.data,
+            #[cfg(feature = "dim3")]
+            JointDescription::SphericalJoint(ref mut j) => &mut j.data,
+            JointDescription::SpringJoint(ref mut j) => &mut j.data,
+        }
+    }
+}
+
+/// A trait used for different constraint types applied to joints,
+/// see [`ImpulseJoint`] and [`MultibodyJoint`].
+pub trait JointConstraint {
+    /// The entity containing the rigid-body used as the first endpoint of this joint.
+    fn parent(&self) -> Entity;
+    /// Access the joint’s description.
+    fn data(&self) -> &JointDescription;
+    /// Access mutably the joint’s description.
+    fn data_mut(&mut self) -> &mut JointDescription;
+}
```

Comment:
I don’t see this trait being actually used anywhere in the public API. Should it be removed?

Reply:
correct